src/qrscanner.py

```diff
@@ -162,15 +166,40 @@ def scan(self, timeout=None):
             return r[:-len(self.EOL)]
         return r
 
-    def update(self):
-        if self.scanning:
-            res = self.uart.read(self.uart.any())
-            if len(res) > 0:
-                self.data += res.decode('utf-8')
-            if self.is_done() and self.callback is not None:
-                data = self.data[:-len(self.EOL)]
-                self.reset()
-                self.callback(data)
+    def update(self, cb_error=None):
+        def process():
+            assert cb_error != None, "callback missing"
+            data = self.data[:-len(self.EOL)]
+            self.reset()
+            qr_anim = QRAnimated.process(data)
+            # is this animated QR code
+            if qr_anim is not None and self.callback_animated_qr is not None:
+                data = qr_anim
+                assert QRAnimated.indx[0] <= QRAnimated.indx[1], "QRAnim: encoding"
+                if QRAnimated.prev_indx == []:
+                    assert QRAnimated.indx[0] == 1, "QRAnim: wrong first code"
+                else:
+                    assert QRAnimated.indx[0] > QRAnimated.prev_indx[0], "QRAnim: wrong order"
+                    assert QRAnimated.indx[1] == QRAnimated.prev_indx[1], "QRAnim: wrong code"
+                self.callback_animated_qr(QRAnimated.indx)
+                QRAnimated.prev_indx = QRAnimated.indx
+                if QRAnimated.indx[0] is not QRAnimated.indx[1]:
+                    return
+            self.callback(data)
+            QRAnimated.clean()
+        try:
+            if self.scanning:
+                res = self.uart.read(self.uart.any())
```

Comment:
This line is problematic for the simulator as there are 2 executions of the tcphost.py::_check() : in self.uart.any() and in uart.read()

If i comment out `self.check()` in read() it works ok:

```
    def read(self, nbytes=None):
        #self._check()
        if nbytes is None or nbytes >= len(self.buf):
            buf = self.buf
            self.buf = b''
        else:
            buf = self.buf[:nbytes]
            self.buf = self.buf[nbytes:]
        return buf
```

I can push the change if you're ok with it.

Reply:
I recently discovered that `uart.read()` will return `None` if there is nothing in the buffer, but returns everything available in the buffer otherwise.
Could we use that then?

```py
res = self.uart.read()
if res is None:
    res = b""
```

Reply:
Fixed